truffleruby/src/main/ruby/core/truffle/interop.rb

```diff
@@ -63,7 +63,17 @@ def self.enumerable(foreign)
     end
 
     class Foreign
+      def hash
+        Foreign.java_hash(self)
+      end
+      
+      def == b
+        self.object_id == b.object_id
```

Comment:
Okay, I've nuked all the changes to Foreign. There's only one place we depend on comparing classes when building the proxies now, so I'll build a specialised case for that.